prov/tcp/src/xnet_av.c

```diff
@@ -97,23 +99,26 @@ static int xnet_mplex_av_insertsym(struct fid_av *av_fid, const char *node,
 	int ret;
 	struct fid_list_entry *item;
 	struct fid_av *subav_fid;
-	fi_addr_t sub_fi_addr;
+	fi_addr_t *sub_fi_addr;
 	struct xnet_mplex_av *av = container_of(av_fid, struct xnet_mplex_av,
 						util_av.av_fid.fid);
 
 	ofi_genlock_lock(&av->lock);
 	ret = ofi_ip_av_insertsym(&av->util_av.av_fid, node, nodecnt,
 				  service, svccnt, fi_addr, flags, context);
-	if (ret)
+	if (ret <= 0)
 		goto out;
+	sub_fi_addr = calloc(1, sizeof(fi_addr_t) * ret);
 	dlist_foreach_container(&av->subav_list, struct fid_list_entry,	item, entry) {
 		subav_fid = container_of(item->fid, struct fid_av, fid);
 		ret = fi_av_insertsym(subav_fid, node, nodecnt, service, svccnt,
-				      &sub_fi_addr, flags, context);
-		if (ret)
+				      sub_fi_addr, flags, context);
+		if (ret <= 0)
 			break;
-		assert(*fi_addr == sub_fi_addr);
+		assert(!fi_addr || memcmp(fi_addr, sub_fi_addr,
+					  sizeof(fi_addr_t) * ret) == 0);
 	}
+	free(sub_fi_addr);
```

Comment:
The error cases are not handled properly. If `ofi_ip_av_insertsym` failed to insert all the addresses (i.e. `ret < nodecnt * svccnt`), `sub_fi_addr` would not have enough space to hold all the addresses returned from `fi_av_insertsym`. In addition, the final return value from this function may not conforming to the API definition.